tests/plugins-by-name.nix

```diff
@@ -0,0 +1,190 @@
+{
+  lib,
+  evaluatedNixvim,
+  linkFarmFromDrvs,
+  runCommandNoCCLocal,
+}:
+let
+  by-name = ../plugins/by-name;
+  options = lib.collect lib.isOption evaluatedNixvim.options;
+
+  # Option namespaces that we allow by-name plugins to declare
+  knownPluginNamespaces = [
+    "colorschemes"
+    "plugins"
+  ];
+
+  # Group by-name children by filetype; "regular", "directory", "symlink" and "unknown".
+  children =
+    let
+      apply =
+        prev: name: type:
+        prev // { ${type} = prev.${type} ++ [ name ]; };
+
+      nil = {
+        regular = [ ];
+        directory = [ ];
+        symlink = [ ];
+        unknown = [ ];
+      };
+    in
+    lib.foldlAttrs apply nil (builtins.readDir by-name);
+
+  # Find plugins by looking for `*.*.enable` options that are declared in `plugins/by-name`
+  by-name-enable-opts =
+    let
+      regex = ''/nix/store/[^/]+/plugins/by-name/(.*)'';
+      optionalPair =
+        opt: file:
+        let
+          result = builtins.match regex file;
+        in
+        lib.optional (result != null) {
+          # Use the file name relative to `plugins/by-name/`
+          name = builtins.head result;
+          # Use only the first two parts of the option location
+          value = lib.genList (builtins.elemAt opt.loc) 2;
+        };
+    in
+    lib.pipe options [
+      (builtins.filter (opt: builtins.length opt.loc == 3 && lib.last opt.loc == "enable"))
+      (builtins.concatMap (opt: (builtins.concatMap (optionalPair opt) opt.declarations)))
+      builtins.listToAttrs
+    ];
+in
+linkFarmFromDrvs "plugins-by-name" [
+  # Ensures all files matching `plugins/by-name/*` are directories
+  (runCommandNoCCLocal "file-types"
+    {
+      __structuredAttrs = true;
+      inherit (children) regular symlink unknown;
+    }
+    ''
+      declare -i errs=0
+
+      showErrs() {
+        type="$1"
+        shift
+
+        if (( $# > 0 )); then
+          ((++errs))
+          echo "Unexpected $type in plugins/by-name ($#):"
+          for f in "$@"; do
+            echo "  - $f"
+          done
+          echo
+        fi
+      }
+
+      showErrs 'symlinks' "''${symlink[@]}"
+      showErrs 'regular files' "''${regular[@]}"
+      showErrs 'unknown-type files' "''${unknown[@]}"
+
+      if (( $errs > 0 )); then
+        exit $errs
```

Comment:
We should always `exit 1`, else if by bad luck we have 256 errors it would return a 0 status code

(Though this is _not_ a big problem, as we would not create an $out file, failing the build anyway)

Reply:
> On POSIX systems the standard exit code is 0 for success and any number from 1 to 255 for anything else

Ah, good catch!